src/access/RequiredPrivilege.cpp

```diff
@@ -0,0 +1,214 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "RequiredPrivilege.h"
+
+namespace {
+
+using namespace chip;
+using namespace chip::Access;
+
+// Privilege override entries are stored in a table per operation (read attribute,
+// write attribute, invoke command, read entry). Entries in each table are sorted
+// by their primary compound key of cluster/endpoint/field, and therefore cannot
+// have duplicate privilege for any such combination. Cluster cannot be invalid,
+// so that is used as a sentinel for empty/unused entries, which are kept at the
+// end of the table. Endpoint and field can be invalid, which means wildcard.
+// Sorting the entries in this way allows a cluster to override specific endpoints
+// or fields with privilege, while still using wildcard to override all the
+// unspecified ones.
+struct PrivilegeOverride
+{
+    ClusterId mCluster;
+    EndpointId mEndpoint;
+    FieldId mField;
+    Privilege mPrivilege;
+
+    PrivilegeOverride() : mCluster(kInvalidClusterId), mEndpoint(kInvalidEndpointId), mField(kInvalidFieldId) {}
+
+    PrivilegeOverride(ClusterId cluster, EndpointId endpoint, FieldId field, Privilege privilege) :
+        mCluster(cluster), mEndpoint(endpoint), mField(field), mPrivilege(privilege)
+    {}
+
+    static_assert(sizeof(FieldId) >= sizeof(AttributeId), "FieldId must be able to hold AttributeId");
+    static_assert(sizeof(FieldId) >= sizeof(CommandId), "FieldId must be able to hold CommandId");
+    static_assert(sizeof(FieldId) >= sizeof(EventId), "FieldId must be able to hold EventId");
+};
+
+PrivilegeOverride
+    privilegeOverrideForReadAttribute[CHIP_CONFIG_ACCESS_CONTROL_MAX_REQUIRED_PRIVILEGE_OVERRIDES_FOR_READ_ATTRIBUTE] = {
```

Comment:
I guess this is not const because of the overrides bits?  Seems to me like the (long-term if we keep this setup) the spec-defined stuff should be const data...

Reply:
Correct. I did envision we could have two tables, one for const, one for non-const, and look at both. But this is what I did in a couple hours.

Reply:
I made it const, someone can write a dynamic part later if we need/want it.